trackers/browser-tracker/test/tab_id_storage.test.ts

```diff
@@ -0,0 +1,102 @@
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import { addTracker, SharedState } from '@snowplow/browser-tracker-core';
+
+describe('Tab identification behaviour', () => {
+  let mockStorage: Record<string, any> = {};
+
+  beforeAll(() => {
+    global.Storage.prototype.setItem = jest.fn((key, value) => {
+      mockStorage[key] = value;
+    });
+    global.Storage.prototype.getItem = jest.fn((key) => mockStorage[key]);
+  });
+
+  beforeEach(() => {
+    mockStorage = {};
+  });
+
+  it('creates a tab identifier and persists across tracking', () => {
+    const state = new SharedState();
+    const tracker = addTracker('sp1', 'sp1', '', '', state) ?? fail('Failed to create tracker');
+
+    tracker.trackPageView();
+    const tabId = tracker.getTabId();
+    expect(tabId).toEqual(window.sessionStorage.getItem('_sp_tab_id'));
+
+    tracker.trackPageView();
+    expect(tabId).toEqual(tracker.getTabId());
+  });
+
+  it('uses a random tab identifier on anonymous tracking', () => {
+    const state = new SharedState();
+    const tracker =
+      addTracker('sp2', 'sp2', '', '', state, { anonymousTracking: true }) ?? fail('Failed to create tracker');
+
+    tracker.trackPageView();
+    expect(window.sessionStorage.getItem('_sp_tab_id')).toBe(undefined);
+    // Every call should generate a random identifier
+    expect(tracker.getTabId()).not.toEqual(tracker.getTabId());
+  });
+
+  it('uses a random tab identifier on stateless strategy', () => {
+    const state = new SharedState();
+    const tracker =
+      addTracker('sp3', 'sp3', '', '', state, { stateStorageStrategy: 'none' }) ?? fail('Failed to create tracker');
+
+    tracker.trackPageView();
+    expect(window.sessionStorage.getItem('_sp_tab_id')).toBe(undefined);
+    // Every call should generate a random identifier
+    expect(tracker.getTabId()).not.toEqual(tracker.getTabId());
+  });
+
+  it('uses the correct stored identifier after disabling anonymous tracking', () => {
+    const state = new SharedState();
+    const tracker =
+      addTracker('sp4', 'sp4', '', '', state, { anonymousTracking: true }) ?? fail('Failed to create tracker');
+
+    tracker.trackPageView();
+    expect(window.sessionStorage.getItem('_sp_tab_id')).toBe(undefined);
+    expect(tracker.getTabId()).not.toEqual(tracker.getTabId());
+
+    tracker.disableAnonymousTracking();
+    tracker.trackPageView();
+    const storedTabId = window.sessionStorage.getItem('_sp_tab_id');
+    const tabId = tracker.getTabId();
+    expect(storedTabId).not.toBe(undefined);
+    expect(tabId).toEqual(storedTabId);
+  });
+
+  afterAll(() => {
```

Comment:
I wouldn't expect to find this at the bottom of the file :) Usually we put them at the top along with the before callbacks. This approach might be better but I think it's more important to be consistent within the project.

Reply:
No particular reason, just the size of the file made it flow to the eye easily. Will move it on top as I agree!